--- a/arch/mips/bmips/setup.c
+++ b/arch/mips/bmips/setup.c
@@ -120,6 +120,11 @@ static void bcm6348_quirks(void)
 	bmips_smp_enabled = 0;
 }
 
+static void bcm6358_quirks(void)
+{
+	bmips_smp_enabled = 0;
+}
+
 static void bcm6368_quirks(void)
 {
 	bcm63xx_fixup_cpu1();
@@ -134,6 +139,7 @@ static const struct bmips_quirk bmips_qu
 	{ "brcm,bcm6338",		&bcm6338_quirks			},
 	{ "brcm,bcm6345",		&bcm6345_quirks			},
 	{ "brcm,bcm6348",		&bcm6348_quirks			},
+	{ "brcm,bcm6358",		&bcm6358_quirks			},
 	{ "brcm,bcm6368",		&bcm6368_quirks			},
 	{ },
 };
