--- a/arch/mips/bcm63xx/dev-flash.c
+++ b/arch/mips/bcm63xx/dev-flash.c
@@ -244,6 +244,10 @@ int __init bcm63xx_flash_register(int nu
 		return spi_register_board_info(bcm63xx_spi_flash_info,
 					ARRAY_SIZE(bcm63xx_spi_flash_info));
 	case BCM63XX_FLASH_TYPE_NAND:
+		/* don't register when already registered through dtb */
+		if (board_of_device_present("nflash"))
+			return 0;
+
 		pr_warn("unsupported NAND flash detected\n");
 		return -ENODEV;
 	default:
